dartagnan/src/main/java/com/dat3m/dartagnan/utils/collections/SetUtil.java

```diff
@@ -84,4 +86,125 @@ public void clear() {
 
         }
     }
+
+    // Manages a distinct union of a modifiable and an immutable set.
+    public static final class GroundedSet<T> extends AbstractSet<T> {
+
+        private final Set<? extends T> ground;
+        private final Set<T> added;
+        private final Set<Object> removed;
```

Comment:
You can make this a `Set<T>`. Notice that you only ever add an element to `removed` if it is contained in `ground`, which guarantees that the removed element must be compatible with type `T`.
Therefore, you can soundly do a type cast in the respective code, i.e., you can do `removed.add((T)o)`.
You can then also add a `@SuppressWarnings("unchecked")` annotation to silence the compiler.